### PR TITLE
SpatialPooler: Stop exempting overlap=0 from stimulusThreshold

### LINKED ISSUE #3287: Winner columns chosen with no overlap with input (local inhibition)
> See discussion of this problem at https://discourse.numenta.org/t/sp-columns-activating-without-input-when-using-topology/1267

Given the condition for selecting winner columns during the local inhibition step here:

https://github.com/numenta/nupic/blob/f1dd3ca676891f7370cf5404b3bbe432ac503f75/src/nupic/research/spatial_pooler.py#L1471

I believe there is a bug that allows columns to be activated that have no overlap with the input vector. When the condition is updated to ensure the column overlaps with the input vector, the proper behavior is observed.

Before:

[](https://www.youtube.com/watch?v=w1t4quc1oNs)

After:

[](https://www.youtube.com/watch?v=BiMZvSbGZKs)

### PR DESCRIPTION
**Pull request by mrcslws**
Fixes #3287

This fixes two bugs. One of these bugs appears in two places.

- Global inhibition: filter out columns with overlap of 0 if `stimulusThreshold > 0`.
- Local inhibiton: filter out columns with overlap of 0 if `stimulusThreshold  > 0`.
- Local inhibition: successfully break ties between columns with overlap of 0. _(only relevant when `stimulusThreshold <= 0`)_

### CODE REVIEW

**Comment by mrcslws** [comment]
This depends on corresponding C++ SP change https://github.com/numenta/nupic.core/pull/1060 to make the SP compatibility tests pass.

**Comment by cogmission** [comment]
Watching this...

**Comment by mrcslws** [comment]
@BoltzmannBrain maybe you could review this? It's adjacent to the tie-breaking code, which is fresh on your mind.

We'll have to merge the [matching C++ change](https://github.com/numenta/nupic.core/pull/1060) before the SP compatibility tests will pass. That's why Travis/AppVeyor are red.

File: src/nupic/research/spatial_pooler.py [hunk 1]
```diff
@@ -1461,16 +1467,23 @@ def _inhibitColumnsLocal(self, overlaps, density):
     @return list with indices of the winning columns
     """
     winners = []
-    addToWinners = max(overlaps)/1000.0
-    overlaps = numpy.array(overlaps, dtype=realDType)
+
+    addToWinners = max(overlaps) / 1000.0
+    if addToWinners == 0:
+      addToWinners = 0.001
```
Comment (BoltzmannBrain) [inline src/nupic/research/spatial_pooler.py]: add comment as to why we're doing this (it's not intuitive)

File: src/nupic/research/spatial_pooler.py [hunk 2]
```diff
@@ -1461,16 +1467,23 @@ def _inhibitColumnsLocal(self, overlaps, density):
...
+
+    tieBrokenOverlaps = numpy.array(overlaps, dtype=realDType)
+
```
Comment (BoltzmannBrain) [inline src/nupic/research/spatial_pooler.py]: better to init `winners` here (closer to its use)

File: src/nupic/research/spatial_pooler.py [hunk 3]
```diff
@@ -1436,10 +1434,18 @@ def _inhibitColumnsGlobal(self, overlaps, density):
 
     # Calculate winners using stable sort algorithm (mergesort)
     # for compatibility with C++
-    winnerIndices = numpy.argsort(overlaps, kind='mergesort')
-    sortedWinnerIndices = winnerIndices[-numActive:][::-1]
+    sortedWinnerIndices = numpy.argsort(overlaps, kind='mergesort')
+
+    # Enforce the stimulus threshold
```
Comment (BoltzmannBrain) [inline src/nupic/research/spatial_pooler.py]: please add this info to the docstring

File: src/nupic/research/spatial_pooler.py [hunk 4]
```diff
@@ -1376,7 +1376,6 @@ def _calculateOverlap(self, inputVector):
     overlaps = numpy.zeros(self._numColumns, dtype=realDType)
     self._connectedSynapses.rightVecSumAtNZ_fast(inputVector.astype(realDType),
                                                  overlaps)
-    overlaps[overlaps < self._stimulusThreshold] = 0
```
Comment (BoltzmannBrain) [inline src/nupic/research/spatial_pooler.py]: please update the docstring (that `stimulusThreshold` logic is not done here)

**Comment by BoltzmannBrain** [comment]
@mrcslws I left some notes about updating docstrings/comments, but otherwise 👍 
Nice unit tests btw 😄

**Comment by mrcslws** [comment]
@BoltzmannBrain good eye, I also just made those changes to the C++ PR.

**Pull request merged**